File: pbmx_ffi/java/io/rmf/pbmx/payloads/PublishKeyPayload.java
package io.rmf.pbmx.payloads;

import io.rmf.pbmx.BlockBuilder;
import io.rmf.pbmx.PublicKey;

public final class PublishKeyPayload extends Payload {
    public PublishKeyPayload(String name, PublicKey key) {
        this.name = name;
        this.key = key;
    }

    @Override
    public void addTo(BlockBuilder builder) {
        builder.publishKey(name, key);
    }

    private final String name;
    private final PublicKey key;
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/payloads/PublishKeyPayload.kt
package io.rmf.pbmx.payloads

import com.sun.jna.Pointer
import com.sun.jna.ptr.LongByReference
import com.sun.jna.ptr.PointerByReference
import io.rmf.pbmx.PublicKey
import io.rmf.pbmx.ffi.FFI

class PublishKeyPayload(handle: Pointer) : Payload(handle) {
    val name: String
    val key: PublicKey

    init {
        val length = LongByReference()
        var r = FFI.pbmx_unwrap_publish_key(this.handle, null, length, null)
        assert(r == 0)

        val outName = ByteArray(length.value.toInt())
        val outKey = PointerByReference()
        r = FFI.pbmx_unwrap_publish_key(this.handle, outName, length, outKey)
        assert(r != 0)
        this.name = outName.toString(Charsets.UTF_8)
        this.key = PublicKey(outKey.value)
    }
}
